src/server/command/worldedit/blueprint.zig

```diff
@@ -0,0 +1,193 @@
+const Vec3i = vec.Vec3i;
+
+const List = main.List;
+const Block = main.blocks.Block;
+const Blueprint = main.blueprint.Blueprint;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+pub const description = "Operate on blueprints.";
+pub const usage = "/blueprint <save FILENAME|delete FILENAME|load FILENAME|list>";
+
+const BlueprintSubCommand = enum {
+	save,
+	delete,
+	load,
+	list,
+	other,
+	empty,
+
+	fn fromString(string: []const u8) BlueprintSubCommand {
+		return std.meta.stringToEnum(BlueprintSubCommand, string) orelse {
+			if(string.len == 0) return .empty;
+			return .other;
+		};
+	}
+};
+
+pub fn execute(args: []const u8, source: *User) void {
+	var argsList = List([]const u8).init(main.stackAllocator);
+
+	var splitIterator = std.mem.splitSequence(u8, args, " ");
+	while(splitIterator.next()) |a| {
+		argsList.append(main.stackAllocator.dupe(u8, a));
+	}
+	defer {
+		for(argsList.items) |arg| {
+			main.stackAllocator.free(arg);
+		}
+		argsList.deinit();
+	}
+
+	if(argsList.items.len < 1) {
+		source.sendMessage("#ff0000Not enough arguments for /blueprint, expected at least 1.", .{});
+		return;
+	}
+	const subcommand = BlueprintSubCommand.fromString(argsList.items[0]);
+	_ = switch(subcommand) {
+		.save => blueprintSave(argsList, source),
+		.delete => blueprintDelete(argsList, source),
+		.load => blueprintLoad(argsList, source),
+		.list => blueprintList(argsList, source),
+		.other => {
+			source.sendMessage("#ff0000Unrecognized subcommand for /blueprint: '{s}'", .{argsList.items[0]});
+		},
+		.empty => {
+			source.sendMessage("#ff0000Missing subcommand for **/blueprint**, usage: {s} ", .{usage});
+		},
+	} catch |err| {
+		std.log.info("Error: {s}", .{@errorName(err)});
+		source.sendMessage("#ff0000Error: {s}", .{@errorName(err)});
+	};
+}
+
+fn blueprintSave(args: List([]const u8), source: *User) !void {
+	if(args.items.len < 2) {
+		source.sendMessage("#ff0000**/blueprint save** requires FILENAME argument.", .{});
+		return;
+	}
+	if(args.items.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint save**. Expected 1 argument, FILENAME.", .{});
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
+
+	if(source.commandData.clipboard) |clipboard| {
+		const storedBlueprint = clipboard.store(main.stackAllocator);
+		defer main.stackAllocator.free(storedBlueprint);
+
+		const fileName: []const u8 = ensureBlueprintExtension(main.stackAllocator, args.items[1]);
+		defer main.stackAllocator.free(fileName);
+
+		var cwd = std.fs.cwd();
+
+		_ = cwd.makeDir("blueprints") catch null;
+
+		var blueprintsDir = try cwd.openDir("blueprints", .{});
+		defer blueprintsDir.close();
+
+		std.log.info("Saving clipboard to blueprint file: {s}", .{fileName});
+		source.sendMessage("#00ff00Saving clipboard to blueprint file: {s}", .{fileName});
+
+		try blueprintsDir.writeFile(.{
+			.sub_path = fileName,
+			.data = storedBlueprint,
+			.flags = .{.lock = .exclusive},
```

Comment:
Please don't silently introduce this into your PR. If this is fixing a relevant issue, then it should be discussed in an issue first.

Reply:
Also quick reminder that main.files exist, so that we don't need to do all these incantations every time we want to read or write a file. e.g. `main.files.openDir` makes `makeDir` obsolete. Calling `.write()` on the result then allows you to write a file directly.

Reply:
Well, It is preventing an issue, not fixing it, I don't like judging synchronization issues as they kick back when you least expect. If the cost of synchronization is low (which it is in this case) It should be done imo. This is relevant only if at any point of existence of the game commands can or will be able to be executed independently for different players. That said, maybe the best approach would be to have sub folders for each player for their own set of blueprints (created when first blueprint is saved), after all even if blueprint files is not corrupted, having someone else overwrite your stuff is not a fun option.